### PR TITLE
bun: install completions

### PR DESCRIPTION
**Pull request by ryanccn**
## Description of changes

Adds zsh and fish completions for [bun](https://bun.sh/).

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by DAlperin — APPROVED** [review]

**Comment by wegank** [comment]
@ofborg build bun bun.passthru.tests

File: pkgs/development/web/bun/default.nix [hunk 1]
```diff
@@ -31,6 +32,18 @@ stdenvNoCC.mkDerivation rec {
 
     runHook postInstall
   '';
+
+  postInstall = ''
```
Comment (wegank) [inline pkgs/development/web/bun/default.nix:36]:
```suggestion
  postPhases = [ "postPatchelf" ];
  postPatchelf = lib.optionalString (stdenvNoCC.buildPlatform.canExecute stdenvNoCC.hostPlatform) ''
```

**Review by wegank — CHANGES_REQUESTED** [review]
`adopt suggestion` -> `bun: adopt suggestion` in the commit message.

File: pkgs/development/web/bun/default.nix [hunk 2]
```diff
@@ -31,6 +32,19 @@ stdenvNoCC.mkDerivation rec {
...
+  postPhases = [ "postPatchelf" ];
+  postPatchelf = lib.optionalString (stdenvNoCC.buildPlatform.canExecute stdenvNoCC.hostPlatform) ''
+    completions_dir=$(mktemp -d)
+
+    SHELL="zsh" $out/bin/bun completions $completions_dir
+    SHELL="fish" $out/bin/bun completions $completions_dir
+
+    installShellCompletion --name bun \
+      --zsh $completions_dir/_bun \
+      --fish $completions_dir/bun.fish
```
Comment (cdmistman) [inline pkgs/development/web/bun/default.nix:45]: bun also has [bash completions](https://github.com/oven-sh/bun/blob/main/completions/bun.bash), can you add those too please?

**Review by cdmistman — CHANGES_REQUESTED** [review]

**Review by cdmistman — APPROVED** [review]
```sh
$ nix shell --refresh github:ryanccn/nixpkgs/bun-completions#bun -c zsh
$ . $(dirname $(which bun))/../share/zsh/site-functions/_bun
$ bun install --ver<tab>
```

works!

**Review by wegank — APPROVED** [review]

**Comment by happysalada** [comment]
ofborg showed a failure on x86_64-darwin, but just confirmed it builds fine.

**Pull request merged**